crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs

```diff
@@ -0,0 +1,221 @@
+use std::{collections::BTreeMap, fmt::Formatter, sync::Arc};
+
+use itertools::Itertools;
+use matrix_sdk::{event_cache::paginator::PaginatorError, Room, SendOutsideWasm, SyncOutsideWasm};
+use matrix_sdk_base::deserialized_responses::SyncTimelineEvent;
+use ruma::{EventId, MilliSecondsSinceUnixEpoch, OwnedEventId};
+use thiserror::Error;
+use tokio::sync::{RwLock, Semaphore};
+
+/// Utility to load the pinned events in a room.
+pub struct PinnedEventsLoader {
+    room: Arc<Box<dyn PinnedEventsRoom>>,
+    max_events_to_load: usize,
+    max_concurrent_requests: usize,
+    cache: PinnedEventCache,
+}
+
+impl PinnedEventsLoader {
+    /// Creates a new `PinnedEventsLoader` instance.
+    pub fn new(room: Box<dyn PinnedEventsRoom>, max_events_to_load: usize) -> Self {
+        Self {
+            room: Arc::new(room),
+            max_events_to_load,
+            max_concurrent_requests: 10,
+            cache: PinnedEventCache {
+                inner: Arc::new(InnerPinnedEventCache { events: Default::default() }),
+            },
+        }
+    }
+
+    /// Loads the pinned events in this room, using the cache first and then
+    /// requesting the event from the homeserver if it couldn't be found.
+    /// This method will perform as many concurrent requests for events as
+    /// `max_concurrent_requests` allows, to avoid overwhelming the server.
+    ///
+    /// It returns a `Result` with either a
+    /// chronologically sorted list of retrieved `SyncTimelineEvent`s  or a
+    /// `PinnedEventsLoaderError`.
+    pub async fn load_events(&self) -> Result<Vec<SyncTimelineEvent>, PinnedEventsLoaderError> {
+        let pinned_event_ids: Vec<OwnedEventId> = self
+            .room
+            .room_pinned_event_ids()
+            .into_iter()
+            .rev()
+            .take(self.max_events_to_load)
+            .rev()
+            .collect();
+
+        let mut loaded_events = Vec::new();
+        let mut event_ids_to_request = Vec::new();
+        for ev_id in pinned_event_ids {
+            if let Some(ev) = self.cache.get(&ev_id).await {
+                loaded_events.push(ev.clone());
+            } else {
+                event_ids_to_request.push(ev_id);
+            }
+        }
+
+        if !event_ids_to_request.is_empty() {
+            let semaphore = Arc::new(Semaphore::new(self.max_concurrent_requests));
```

Comment:
Can you extract the logic for performing-things-concurrently-with-a-limit into a separate function, and then pass in the list of inputs (event IDs) and a closure for the thing to do? Don't worry if not.

Reply:
It seems like I can't because I'd have to make the closure async and that's unstable.

Reply:
Normally when you want an async closure you can get away with a closure that returns a Future by doing:

```
|args| async { body }
```

Not sure here.

Reply:
I think I'll try doing this in a separate PR, since the team implementing the feature is a bit blocked without it merged.

Reply:
Yep, fine - thanks.